### PR TITLE
Add type hints to methods of interface, including arguments

### PR DESCRIPTION
**Pull request by mwcraig**
Pretty sure I did this right, but this is my first genuine attempt at type hinting, so 🤷‍♂️ 

Related to #167

### CODE REVIEW

**Comment by mwcraig** [comment]
Test failure here is because I'm using the `Union` notation introduced in `Python 3.10`, i.e. I'm doing `str | None` as opposed to `Union[str, None]` or `Optional[str`. 

Are we ok with making 3.10 the minimum version?

**Comment by pllim** [comment]
> Are we ok with making 3.10 the minimum version?

Fine by me. I think all this means is downstream needs that version to check against Protocol, but doesn't mean they themselves need to support that minversion, right?

**Comment by mwcraig** [comment]
I think that is correct. If we split the API out into its own package then users never need to install the API package; it could simply be a testing dependency for the downstream packages.

**Comment by pllim** [comment]
In that case, bumping to Python 3.10 is not controversial. :shipit:

**Comment by mwcraig** [comment]
I think this is ready for merge -- I've updated the minimum python version 3.10 so that we can use the `|` operator in type hints.

File: .readthedocs.yml [hunk 1]
```diff
@@ -9,7 +9,7 @@ sphinx:
 
 # Set the version of Python and requirements required to build your docs
 python:
-  version: 3.8
+  version: 3.10
```
Comment (pllim) [inline .readthedocs.yml:12]: You have to use the new build syntax. See https://docs.readthedocs.io/en/stable/config-file/v2.html

**Comment by mwcraig** [comment]
All tests are passing now -- thanks for the RTD fix, @pllim!

**Review by pllim — APPROVED** [review]
Looks uncontroversial enough. Thanks!

**Pull request merged**